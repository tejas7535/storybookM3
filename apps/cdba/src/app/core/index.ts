export { CoreModule } from './core.module';
export { RoleGuard } from './guards';
