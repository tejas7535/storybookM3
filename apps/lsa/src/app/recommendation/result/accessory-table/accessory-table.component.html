@if (!showEmptyState) {
  <table
    class="accessory-table"
    [formGroup]="tableFormGroup"
    *transloco="let t; read: 'recommendation.result'"
  >
    <thead class="text-left">
      <tr class="[&>th]:py-2 [&>th]:px-2 border-b-2 border-border">
        <th>{{ t('product') }}</th>
        <th>
          <span class="sr-only">{{ t('availability') }}</span>
        </th>
        @if (isBusinessUser) {
          <th class="text-center">{{ t('price') }}</th>
        }
        <th class="w-11 md:w-20">{{ t('quantity') }}</th>
        <th>
          <span class="sr-only">{{ t('actions') }}</span>
        </th>
      </tr>
    </thead>
    <tbody>
      @for (group of accGroups | SortedAccessoryList; track group.key) {
        <tr class="accessory-group">
          <th class="text-left">
            {{ t('category.' + group.value.groupTitle) }}
          </th>
          <th>&nbsp;</th>

          @if (isBusinessUser) {
            <th class="text-center">
              @if (currency) {
                <span class="inline-block">
                  {{
                    tableGroupStates[group.key].totalNetPrice
                      | currency: currency
                      | lsaCustomCurrency
                  }}
                </span>
              } @else {
                <span class="inline-block">
                  {{
                    tableGroupStates[group.key].totalNetPrice | number: '1.2-2'
                  }}
                </span>
              }
            </th>
          }
          <th>{{ tableGroupStates[group.key].totalQty }} pcs</th>
          <th class="w-12">
            <button
              class="flex w-full justify-center"
              (click)="onToggleGroupClick(group.key)"
            >
              <mat-icon aria-hidden="true">{{
                tableGroupStates[group.key].isOpen
                  ? 'keyboard_arrow_up'
                  : 'keyboard_arrow_down'
              }}</mat-icon>
              <div class="sr-only">
                {{ tableGroupStates[group.key].isOpen ? 'Collapse' : 'Expand' }}
                group {{ group.value.groupTitle }}
              </div>
            </button>
          </th>
        </tr>
        @if (tableGroupStates[group.key].isOpen) {
          @for (item of group.value.items; track item.fifteen_digit) {
            <tr class="accessory-item" [formGroupName]="group.key">
              <td class="pl-1 last-of-type:pr-1">
                <div
                  class="flex items-center flex-col md:flex-row space-y-1 md:space-y-0"
                >
                  <img
                    [src]="
                      item.product_image ? item.product_image : imagePlaceholder
                    "
                    class="mr-1 aspect-square object-scale-down w-16 h-16 flex-shrink-0"
                    alt=""
                  />
                  <div class="flex flex-col justify-between">
                    <span class="text-title-small text-medium-emphasis">{{
                      item.designation
                    }}</span>
                    <span class="text-low-emphasis text-body-small mt-1">{{
                      item.description
                    }}</span>
                    <span class="text-low-emphasis mt-2 text-[10px]/[18px]"
                      >Schaeffler ID {{ item.fifteen_digit }}</span
                    >
                  </div>
                </div>
              </td>
              <td>
                @if (isBusinessUser) {
                  @if (item?.availability === true) {
                    <div class="flex flex-col items-center space-y-1">
                      <div
                        class="aspect-square w-3 h-3 rounded-full bg-primary"
                      >
                        <div class="sr-only">Availability indicator green</div>
                      </div>
                      <span
                        class="text-body-small text-medium-emphasis text-center"
                        >{{ t('inStock') }}</span
                      >
                    </div>
                  }

                  @if (item?.availability === false) {
                    <div class="flex flex-col items-center space-y-1">
                      <div
                        class="aspect-square w-3 h-3 rounded-full bg-[#c1c0c0]"
                      >
                        <div class="sr-only">
                          Availability indicator unavailable
                        </div>
                      </div>
                      <span
                        class="text-body-small text-medium-emphasis text-center"
                        >{{ t('notInStock') }}</span
                      >
                    </div>
                  }

                  @if (!item.isPriceAndAvailabilityUpdated) {
                    <div class="rounded-md p-4 max-w-sm w-full mx-auto">
                      <div class="animate-pulse flex space-x-4">
                        <div class="flex-1 space-y-6 py-1">
                          <div class="space-y-3">
                            <div class="flex flex-col gap-4">
                              <div
                                class="h-2 bg-surface-container-highest rounded col-span-2"
                              ></div>
                              <div
                                class="h-2 bg-surface-container-highest rounded col-span-2"
                              ></div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  }

                  @if (
                    item.isPriceAndAvailabilityUpdated &&
                    item?.availability === undefined
                  ) {
                    <div class="flex flex-col items-center space-y-1">-</div>
                  }
                }
              </td>
              @if (isBusinessUser) {
                <td>
                  <div class="text-center flex flex-col justify-center h-full">
                    @if (item?.price !== null && item?.price !== undefined) {
                      <span class="text-body-small text-medium inline-block">
                        {{
                          item?.price | currency: currency | lsaCustomCurrency
                        }}
                      </span>
                      <span
                        class="text-body-small mt-1 text=[12px] text-medium-emphasis"
                        >{{ t('perPiece') }}</span
                      >
                    } @else if (!item.isPriceAndAvailabilityUpdated) {
                      <div
                        class="text-center flex flex-col justify-center h-full"
                      >
                        <div class="rounded-md p-4 max-w-sm w-full mx-auto">
                          <div class="animate-pulse flex space-x-4">
                            <div class="flex-1 space-y-6 py-1">
                              <div class="space-y-3">
                                <div class="flex flex-col gap-4">
                                  <div
                                    class="h-2 bg-surface-container-highest rounded col-span-2"
                                  ></div>
                                  <div
                                    class="h-2 bg-surface-container-highest rounded col-span-2"
                                  ></div>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    } @else {
                      <span class="text-body-small text-medium inline-block">
                        -
                      </span>
                    }
                  </div>
                </td>
              }

              <td colspan="2" class="px-2">
                <div>
                  <span class="text-body-small">{{ t('pieces') }}</span>
                  <div class="flex flex-col md:flex-row items-center">
                    <input
                      type="number"
                      min="0"
                      [formControlName]="item.fifteen_digit"
                      aria-label="Quantity"
                      class="border-b-2 text-center focus:border-primary active:border-primary border-border outline-none w-full md:w-8"
                    />
                    <div class="flex mt-1 md:mt-0">
                      <button
                        (click)="changeQty(group.key, item.fifteen_digit, '-')"
                        mat-icon-button
                        class="icon-button-primary"
                        [disabled]="
                          !minusButtonEnabled(group.key, item.fifteen_digit)
                        "
                      >
                        <mat-icon>remove</mat-icon>
                        <div class="sr-only">{{ t('increaseOne') }}</div>
                      </button>

                      <button
                        (click)="changeQty(group.key, item.fifteen_digit, '+')"
                        mat-icon-button
                        class="icon-button-primary"
                      >
                        <mat-icon>add</mat-icon>
                        <div class="sr-only">{{ t('decreaseOne') }}</div>
                      </button>
                    </div>
                  </div>
                </div>
              </td>
            </tr>
          }
        }

        <tr class="accessory-group-spacing">
          <td colspan="5"></td>
        </tr>
      }
    </tbody>
    <tfoot>
      <tr>
        <th scope="row" colspan="2">
          <span class="sr-only">{{ t('configSummary') }}</span>
        </th>
        @if (!isNaN(tableSummaryState.totalPrice) && isBusinessUser) {
          <td>
            <div class="flex flex-col px-2 text-center">
              <span class="font-semibold inline-block">
                {{
                  tableSummaryState.totalPrice
                    | currency: currency
                    | lsaCustomCurrency
                }}</span
              >
              <span class="font-light text-body-small">{{
                t('totalPrice')
              }}</span>
            </div>
          </td>
        } @else {
          <td></td>
        }
        <td>
          <div class="flex flex-col px-2">
            <span class="font-semibold">{{ tableSummaryState.totalQty }}</span>
            <span class="font-light text-body-small">{{
              t('totalPieces')
            }}</span>
          </div>
        </td>
      </tr>
    </tfoot>
  </table>
} @else {
  <div>No accessory configuration found</div>
}
