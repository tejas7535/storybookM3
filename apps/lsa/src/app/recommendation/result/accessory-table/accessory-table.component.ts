/* eslint-disable @typescript-eslint/member-ordering */
import { JsonPipe, KeyValuePipe } from '@angular/common';
import {
  Component,
  Input,
  OnChanges,
  OnDestroy,
  SimpleChanges,
} from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';

import { Subject, takeUntil } from 'rxjs';

import { Accessory } from '@lsa/shared/models';
import { PushPipe } from '@ngrx/component';

import { SharedTranslocoModule } from '@schaeffler/transloco';

import {
  generateFormGroup,
  transformAccessories,
} from './accessory-table.helper';
import {
  AccessoryTable,
  AccessoryTableFormGroup,
  TableGroupState,
} from './accessory-table.model';

@Component({
  selector: 'lsa-accessory-table',
  templateUrl: './accessory-table.component.html',
  standalone: true,
  imports: [
    MatIconModule,
    KeyValuePipe,
    FormsModule,
    ReactiveFormsModule,
    MatButtonModule,
    PushPipe,
    JsonPipe,
    SharedTranslocoModule,
  ],
})
export class AccessoryTableComponent implements OnChanges, OnDestroy {
  @Input()
  public readonly accessories: Accessory[];

  public tableFormGroup!: AccessoryTableFormGroup;

  public accGroups: AccessoryTable;
  public tableGroupStates: { [key: string]: TableGroupState };
  public tableSummaryState: { totalQty: number; totalPrice?: number } = {
    totalQty: 0,
  };

  private readonly destroy$ = new Subject<void>();
  private readonly formUpdate$ = new Subject<void>();

  public showEmptyState = true;

  ngOnChanges(changes: SimpleChanges): void {
    if (changes.accessories.currentValue) {
      this.formUpdate$.next();
      this.accGroups = {};
      this.tableGroupStates = {};
      this.tableSummaryState = { totalQty: 0 };
      this.generateAccessoriesForInput();
      this.showEmptyState = false;
    } else {
      this.showEmptyState = true;
    }
  }

  ngOnDestroy(): void {
    this.destroy$.next();
  }

  generateAccessoriesForInput(): void {
    this.accGroups = transformAccessories(this.accessories);
    this.tableFormGroup = generateFormGroup(this.accGroups);

    this.tableFormGroup.valueChanges
      .pipe(takeUntil(this.formUpdate$))
      .subscribe((newValue) => {
        this.tableSummaryState.totalQty = 0;
        for (const [key, value] of Object.entries(newValue)) {
          const qty = Object.values(value).reduce(
            (curr, next) => curr + next,
            0
          );
          // TODO: update price
          this.tableSummaryState.totalQty += qty;
          this.tableGroupStates[key].totalQty = qty;
        }
      });

    const states: { [key: string]: TableGroupState } = {};

    Object.entries(this.accGroups).forEach(([group, value]) => {
      const totalQty = value.items.reduce(
        (prev, current) => prev + current.qty,
        0
      );
      const totalNetPrice = value.items
        .filter((acc) => acc.price)
        .reduce((prev, current) => prev + current.price, 0);

      states[group] = { isOpen: true, totalQty, totalNetPrice };

      this.tableSummaryState.totalQty += totalQty;
      this.tableSummaryState.totalPrice += totalNetPrice;
    });

    this.tableGroupStates = states;
  }

  onToggleGroupClick(groupName: string) {
    if (this.tableGroupStates[groupName]) {
      this.tableGroupStates[groupName].isOpen =
        !this.tableGroupStates[groupName].isOpen;
    }
  }

  changeQty(groupName: string, controlName: string, direction: '-' | '+') {
    const originalValue = this.tableFormGroup
      .get(groupName)
      .get(controlName).value;
    this.tableFormGroup
      .get(groupName)
      .get(controlName)
      .setValue(direction === '+' ? originalValue + 1 : originalValue - 1);
  }

  minusButtonEnabled(groupName: string, formControlName: string): boolean {
    return this.tableFormGroup.get(groupName).get(formControlName).value >= 1;
  }

  isNaN(test: number | any): boolean {
    return Number.isNaN(test);
  }

  shouldShowPriceColumn(): boolean {
    return Object.values(this.tableGroupStates).some(
      (state) => !!state.totalNetPrice
    );
  }
}
