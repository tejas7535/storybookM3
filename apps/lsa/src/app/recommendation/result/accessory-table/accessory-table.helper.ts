import { FormControl, FormGroup, Validators } from '@angular/forms';

import { Accessory } from '@lsa/shared/models';

import {
  AccessoryTable,
  AccessoryTableFormGroup,
  AccessoryTableGroup,
} from './accessory-table.model';

export function transformAccessories(
  accessories: Accessory[],
  priorityLookup: Map<string, number>
): {
  [key: string]: AccessoryTableGroup;
} {
  const groups = new Map<string, Accessory[]>();
  for (const acc of accessories) {
    const className = `${acc.class}`;
    if (!groups.has(className)) {
      groups.set(className, []);
    }
    const expansion = [acc, ...groups.get(className)];
    groups.set(className, expansion);
  }

  const tableGroups = {} as { [key: string]: AccessoryTableGroup };
  for (const [key, items] of groups.entries()) {
    const classId = items[0].class_id;
    const keyFromMap = priorityLookup.get(classId);
    tableGroups[key] = {
      groupTitle: toCamelCase(key),
      groupClassId: classId,
      groupClassPriority: keyFromMap || 0,
      items: items.sort((a) => (a.is_recommendation ? -1 : 0)),
    };
  }

  return tableGroups;
}

export function generateFormGroup(
  tableData: AccessoryTable
): AccessoryTableFormGroup {
  const newFormGroup: { [key: string]: FormGroup } = {};
  for (const [groupKey, value] of Object.entries(tableData)) {
    const group: { [key: string]: FormControl } = {};
    for (const item of value.items) {
      group[item.fifteen_digit] = new FormControl(item.qty, [
        Validators.required,
        Validators.min(0),
      ]);
    }
    newFormGroup[groupKey] = new FormGroup(group);
  }

  return new FormGroup(newFormGroup);
}

export function toCamelCase(str: string): string {
  return str
    .split(/[\s-_]+/) // Split by spaces, hyphens, or underscores
    .map((word, index) => {
      if (index === 0) {
        return word.toLowerCase();
      }

      return word.charAt(0).toUpperCase() + word.slice(1).toLowerCase();
    })
    .join('');
}
