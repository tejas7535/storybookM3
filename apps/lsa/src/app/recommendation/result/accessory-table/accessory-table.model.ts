import { FormControl, FormGroup } from '@angular/forms';

import { Accessory } from '@lsa/shared/models';

export type TableItem = Accessory;

export interface CartSummary {
  totalNetPrice?: number;
  totalPieceCount?: number;
}

export interface AccessoryTableGroup {
  groupTitle: string;
  items: TableItem[];
}

export interface AccessoryTable {
  [key: string]: AccessoryTableGroup;
}

export interface TableGroupState {
  isOpen: boolean;
  totalQty: number;
  totalNetPrice?: number;
}

export type AccessoryTableFormGroup = FormGroup<{
  [key: string]: FormGroup<{ [key: string]: FormControl<number> }>;
}>;
