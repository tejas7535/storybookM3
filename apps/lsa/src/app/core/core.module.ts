import {
  provideHttpClient,
  withInterceptorsFromDi,
} from '@angular/common/http';
import { NgModule } from '@angular/core';

import { TranslocoService } from '@jsverse/transloco';
import { environment } from '@lsa/environments/environment';
import {
  AVAILABLE_LANGUAGES,
  FALLBACK_LANGUAGE,
  LANGUAGE_STORAGE_KEY,
} from '@lsa/shared/constants/language';

import { SharedTranslocoModule } from '@schaeffler/transloco';

import deJson from '../../assets/i18n/de.json';
import enJson from '../../assets/i18n/en.json';

const assetPath = environment.assetsPath;

@NgModule({
  exports: [SharedTranslocoModule],
  imports: [
    SharedTranslocoModule.forRoot(
      environment.production,
      AVAILABLE_LANGUAGES,
      undefined, // default -> undefined would lead to browser detection
      FALLBACK_LANGUAGE.id,
      LANGUAGE_STORAGE_KEY,
      true,
      !environment.localDev,
      undefined,
      `${assetPath}/i18n`
    ),
  ],
  providers: [provideHttpClient(withInterceptorsFromDi())],
})
export class CoreModule {
  public constructor(private readonly translocoService: TranslocoService) {
    this.translocoService.setTranslation(enJson, 'en');
    this.translocoService.setTranslation(deJson, 'de');
  }
}
