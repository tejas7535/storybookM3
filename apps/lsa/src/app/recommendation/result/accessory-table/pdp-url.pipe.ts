import { Pipe, PipeTransform } from '@angular/core';

import { MediasURLBuilderService } from '@lsa/core/services/medias-url-builder.service';

@Pipe({
  standalone: true,
  name: 'pdpurl',
})
export class PdpPageUrlPipe implements PipeTransform {
  constructor(private readonly mediasUrlBuilder: MediasURLBuilderService) {}

  transform(value: string) {
    return this.mediasUrlBuilder.getMediasPDPUrl(value);
  }
}
