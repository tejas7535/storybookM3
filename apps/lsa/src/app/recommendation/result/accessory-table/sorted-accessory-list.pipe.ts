import { Pipe, PipeTransform } from '@angular/core';

import { AccessoryTableGroup } from './accessory-table.model';

@Pipe({
  name: 'SortedAccessoryList',
})
export class SortedAccessoryListPipe implements PipeTransform {
  transform(value: { [key: string]: AccessoryTableGroup }) {
    const keymap = Object.entries(value)
      .sort((a, b) => b[1].groupClassPriority - a[1].groupClassPriority)
      .map((item) => ({ key: item[0], value: item[1] }));

    return keymap;
  }
}
