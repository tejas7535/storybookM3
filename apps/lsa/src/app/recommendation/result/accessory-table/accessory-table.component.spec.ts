import { SimpleChange, SimpleChanges } from '@angular/core';
import { FormControl, FormGroup } from '@angular/forms';

import { Subject } from 'rxjs';

import { PDFGeneratorService } from '@lsa/core/services/pdf-generation/pdf-generator.service';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';

import { provideTranslocoTestingModule } from '@schaeffler/transloco/testing';

import { AccessoryTableComponent } from './accessory-table.component';
import {
  generateFormGroup,
  transformAccessories,
} from './accessory-table.helper';
import { AccessoryTableGroup, TableItem } from './accessory-table.model';

jest.mock('./accessory-table.helper', () => ({
  generateFormGroup: jest.fn(),
  transformAccessories: jest.fn(),
  valueChanges: jest.fn(),
}));

describe('AccessoryTableComponent', () => {
  let spectator: Spectator<AccessoryTableComponent>;
  let component: AccessoryTableComponent;

  const createComponent = createComponentFactory({
    component: AccessoryTableComponent,
    imports: [provideTranslocoTestingModule({ en: {} })],
    providers: [
      {
        provide: PDFGeneratorService,
        useValue: {
          setFormData: jest.fn(),
        },
      },
    ],
  });

  beforeEach(() => {
    spectator = createComponent();
    component = spectator.component;
  });

  it('should create the component', () => {
    expect(spectator.component).toBeTruthy();
  });

  describe('ngOnChanges', () => {
    const className = 'Adapter for lubricator';
    const firstFifteenDigit = '789666050000010';
    const secondFifteenDigit = '123466050000010';
    let mockFormGroup: FormGroup;
    let mockBaseTableGroups: { [key: string]: AccessoryTableGroup };
    let changes: SimpleChanges;

    beforeEach(() => {
      mockFormGroup = new FormGroup({
        'Adapter for lubricator': new FormGroup({
          [firstFifteenDigit]: new FormControl<number>(1),
          [secondFifteenDigit]: new FormControl<number>(2),
        }),
      });

      (generateFormGroup as jest.Mock).mockReturnValue(mockFormGroup);

      mockBaseTableGroups = {
        'Adapter for lubricator': {
          groupTitle: className,
          groupClassId: '1',
          items: [
            {
              pim_code: '123456',
              fifteen_digit: firstFifteenDigit,
              class: 'Adapter',
            } as Partial<TableItem> as TableItem,
            {
              pim_code: '7890321',
              fifteen_digit: secondFifteenDigit,
              class: 'Adapter',
            } as Partial<TableItem> as TableItem,
          ],
        },
      };

      const accessories = [
        { id: 1, name: 'Accessory 1', class: 'Adapter for lubricator' },
      ];
      changes = {
        accessories: new SimpleChange(undefined, accessories, true),
      };
    });

    it('should provide state with pricing information', () => {
      const mockTableGroups = {
        'Adapter for lubricator': {
          groupTitle: className,
          items: [
            {
              ...mockBaseTableGroups[className].items[0],
              qty: 0,
              price: 0.51,
              currency: '$',
            },
            {
              ...mockBaseTableGroups[className].items[1],
              qty: 1,
              price: 0.68,
              currency: '$',
            },
          ],
        },
      };

      (transformAccessories as jest.Mock).mockReturnValue(mockTableGroups);

      jest.spyOn(component['formUpdate$'], 'next');
      jest.spyOn(spectator.component, 'generateAccessoriesForInput');

      spectator.component.ngOnChanges(changes);

      expect(spectator.component['formUpdate$'].next).toHaveBeenCalled();
      expect(spectator.component.accGroups).toEqual(mockTableGroups);
      expect(spectator.component.tableGroupStates).toMatchSnapshot();

      expect(
        spectator.component.generateAccessoriesForInput
      ).toHaveBeenCalled();
      expect(spectator.component.showEmptyState).toBe(false);
    });

    it('should provide pricing information with euro', () => {
      const mockItems = [
        {
          ...mockBaseTableGroups[className].items[0],
          qty: 0,
          price: 0.51,
          currency: '€',
        },
        {
          ...mockBaseTableGroups[className].items[1],
          qty: 1,
          price: 0.68,
          currency: '€',
        },
      ];

      const mockTableGroups = {
        [className]: {
          groupTitle: className,
          items: mockItems,
        },
      };

      (transformAccessories as jest.Mock).mockReturnValue(mockTableGroups);

      spectator.component.ngOnChanges(changes);

      expect(spectator.component.tableGroupStates).toMatchSnapshot();
    });

    it('should set showEmptyState to true when accessories is null', () => {
      const newChanges: SimpleChanges = {
        accessories: new SimpleChange(
          [{ id: 1, name: 'Accessory 1' }],
          undefined,
          true
        ),
      };

      spectator.component.ngOnChanges(newChanges);

      expect(spectator.component.showEmptyState).toBe(true);
    });

    describe('when values are changing', () => {
      beforeEach(() => {
        const mockItems = [
          {
            ...mockBaseTableGroups[className].items[0],
            qty: 0,
            price: 0.51,
            currency: '€',
          },
          {
            ...mockBaseTableGroups[className].items[1],
            qty: 1,
            price: 0.68,
            currency: '€',
          },
        ];

        const mockTableGroups = {
          [className]: {
            groupTitle: className,
            items: mockItems,
          },
        };

        (transformAccessories as jest.Mock).mockReturnValue(mockTableGroups);

        spectator.component.ngOnChanges(changes);
      });

      it('should recalculates the totalQty and totalPrice', () => {
        mockFormGroup.patchValue({
          [className]: {
            [firstFifteenDigit]: 10,
          },
        });

        mockFormGroup.updateValueAndValidity();

        expect(spectator.component.tableGroupStates).toMatchSnapshot();
      });
    });
  });

  describe('ngOnDestroy', () => {
    it('should complete the observable', () => {
      const destroy$ = spectator.component['destroy$'] as Subject<void>;
      const nextSpy = jest.spyOn(destroy$, 'next');
      const completeSpy = jest.spyOn(destroy$, 'complete');

      component.ngOnDestroy();

      expect(nextSpy).toHaveBeenCalled();
      expect(completeSpy).toHaveBeenCalled();
    });
  });

  describe('should show price columnt', () => {
    it('should return true if any state has a totalNetPrice', () => {
      component.tableGroupStates = {
        group1: { isOpen: true, totalQty: 0, totalNetPrice: 100 },
        group2: { isOpen: true, totalQty: 0, totalNetPrice: 0 },
      };
      expect(component.shouldShowPriceColumn()).toBe(true);
    });
    it('should return false if tableGroupStates is empty', () => {
      component.tableGroupStates = {};
      expect(component.shouldShowPriceColumn()).toBe(false);
    });
  });

  describe('isNaN', () => {
    it('should return true for NaN', () => {
      expect(component.isNaN(Number.NaN)).toBe(true);
    });

    it('should return false for a number', () => {
      expect(component.isNaN(123)).toBe(false);
    });

    it('should return false for a number string', () => {
      expect(component.isNaN('123')).toBe(false);
    });
  });

  describe('when setting priceAvailabiltyResponses', () => {
    let mockFormGroup: FormGroup;
    const firstFifteenDigit = '789666050000010';
    const secondFifteenDigit = '123466050000010';
    const className = 'Adapter for lubricator';
    let mockBaseTableGroups: { [key: string]: AccessoryTableGroup };

    beforeEach(() => {
      mockBaseTableGroups = {
        'Adapter for lubricator': {
          groupTitle: className,
          groupClassId: '1',
          items: [
            {
              pim_code: '123456',
              fifteen_digit: firstFifteenDigit,
              class: 'Adapter',
            } as Partial<TableItem> as TableItem,
            {
              pim_code: '7890321',
              fifteen_digit: secondFifteenDigit,
              class: 'Adapter',
            } as Partial<TableItem> as TableItem,
          ],
        },
      };

      const mockTableGroups = {
        'Adapter for lubricator': {
          groupTitle: className,
          items: [
            {
              ...mockBaseTableGroups[className].items[0],
            },
            {
              ...mockBaseTableGroups[className].items[1],
            },
          ],
        },
      };

      (transformAccessories as jest.Mock).mockReturnValue(mockTableGroups);

      mockFormGroup = new FormGroup({
        'Adapter for lubricator': new FormGroup({
          [firstFifteenDigit]: new FormControl<number>(1),
          [secondFifteenDigit]: new FormControl<number>(2),
        }),
      });

      (generateFormGroup as jest.Mock).mockReturnValue(mockFormGroup);

      spectator.setInput('accessories', [
        {
          id: 1,
          matnr: firstFifteenDigit,
          pim_code: '123456',
          name: 'Accessory 1',
          class: 'Adapter for lubricator',
        },
      ]);

      spectator.setInput('priceAndAvailabilityResponses', {
        '123456': { available: true, price: 100, currency: 'USD' },
        '7890321': { available: false, currency: 'USD' },
      });
    });

    it('should update price and availability', () => {
      spectator.detectChanges();

      expect(component.priceAndAvailabilityResponses).toMatchSnapshot();
    });

    it('should update accessory with price and availability', () => {
      spectator.detectChanges();

      expect(component.accessories).toMatchSnapshot();
    });
  });
});
