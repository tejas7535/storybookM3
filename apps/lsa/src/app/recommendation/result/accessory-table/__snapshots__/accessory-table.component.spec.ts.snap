// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccessoryTableComponent ngOnChanges should provide pricing information with euro 1`] = `
Object {
  "Adapter for lubricator": Object {
    "isOpen": true,
    "totalNetPrice": 0.68,
    "totalQty": 1,
  },
}
`;

exports[`AccessoryTableComponent ngOnChanges should provide state with pricing information 1`] = `
Object {
  "Adapter for lubricator": Object {
    "isOpen": true,
    "totalNetPrice": 0.68,
    "totalQty": 1,
  },
}
`;

exports[`AccessoryTableComponent ngOnChanges when values are changing should recalculates the totalQty and totalPrice 1`] = `
Object {
  "Adapter for lubricator": Object {
    "isOpen": true,
    "totalNetPrice": 6.46,
    "totalQty": 12,
  },
}
`;
